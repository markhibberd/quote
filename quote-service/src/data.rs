
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UserAccount {
    pub key: Key,
    pub email: String,
}


#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QuoteFile {
    pub name: String,
    pub access: Permission,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token(pub String);

impl Token {
    pub fn as_string(self) -> String {
        let Token(v) = self; v
    }

    pub fn to_str(&self) -> &str {
        let Token(v) = self; &v
    }
}


#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Keyed<A> {
    pub key: Key,
    pub value: A,
}

impl<A> Keyed<A> {
    pub fn map<B, F: FnOnce(A) -> B>(self, f: F) -> Keyed<B> {
        let Keyed { key, value } = self;
        let value = f(value);
        Keyed { key, value }
    }
}
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Key(pub i64);

impl Key {
    pub fn to_i64(&self) -> i64 {
        let Key(v) = self; *v
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Permission {
    Manage,
    Write,
    Read,
}

impl Permission {
    pub fn to_i32(&self) -> i32 {
       match self {
           Permission::Manage => 2,
           Permission::Write => 1,
           Permission::Read => 0,
       }
    }

    pub fn from_i32(v: i32) -> Permission {
        match v {
            2 => Permission::Manage,
            1 => Permission::Write,
            _ => Permission::Read,
        }
    }
}

#[cfg(test)]
mod test {
    use crate::data::Permission;

    #[test]
    fn permission_i32_symmetry() {
        assert_eq!(Permission::Manage, Permission::from_i32(Permission::Manage.to_i32()));
        assert_eq!(Permission::Write, Permission::from_i32(Permission::Write.to_i32()));
        assert_eq!(Permission::Read, Permission::from_i32(Permission::Read.to_i32()));
    }

    #[test]
    fn permission_i32_out_of_bounds_safe_fallback() {
        assert_eq!(Permission::Read, Permission::from_i32(0));
        assert_eq!(Permission::Read, Permission::from_i32(-1));
        assert_eq!(Permission::Read, Permission::from_i32(3));
        assert_eq!(Permission::Read, Permission::from_i32(123456));
    }
}
