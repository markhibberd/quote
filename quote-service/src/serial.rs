use crate::data;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct UserCreateRequest {
    pub email: String,
    pub password: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct UserCreateResponse {
    pub email: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct SessionCreateRequest {
    pub email: String,
    pub password: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct SessionCreateResponse {
    pub token: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteFileCreateRequest {
    pub name: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteFileCreateResponse {
    #[serde(flatten)]
    pub file: QuoteFile,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteFileListResponse {
    pub files: Vec<QuoteFile>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteListResponse {
    #[serde(flatten)]
    pub file: QuoteFile,
    pub quotes: Vec<Quote>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteGetResponse {
    #[serde(flatten)]
    pub quote: Quote,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum Permission {
    Manage,
    Write,
    Read,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct QuoteFile {
    pub id: i64,
    pub name: String,
    pub access: Permission,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct Quote {
    pub id: i64,
    pub content: String,
}

impl From<Permission> for data::Permission {
    fn from(v: Permission) -> data::Permission {
        match v {
            Permission::Manage => data::Permission::Manage,
            Permission::Write => data::Permission::Write,
            Permission::Read => data::Permission::Read,
        }
    }
}

impl From<data::Permission> for Permission {
    fn from(v: data::Permission) -> Permission {
        match v {
            data::Permission::Manage => Permission::Manage,
            data::Permission::Write => Permission::Write,
            data::Permission::Read => Permission::Read,
        }
    }
}


impl From<QuoteFile> for data::Keyed<data::QuoteFile> {
    fn from(v: QuoteFile) -> data::Keyed<data::QuoteFile> {
        data::Keyed {
            key: data::Key(v.id),
            value: data::QuoteFile {
                name: v.name,
                access: v.access.into(),
            },
        }
    }
}

impl From<data::Keyed<data::QuoteFile>> for QuoteFile {
    fn from(v: data::Keyed<data::QuoteFile>) -> QuoteFile {
        QuoteFile {
            id: v.key.to_i64(),
            name: v.value.name,
            access: v.value.access.into(),
        }
    }
}


#[cfg(test)]
mod test {
    use crate::serial::{
        QuoteFile,
        Permission,
        SessionCreateRequest,
        SessionCreateResponse,
        UserCreateRequest,
        UserCreateResponse,
        QuoteFileCreateRequest,
        QuoteFileCreateResponse,
        QuoteFileListResponse,
    };

    use insta::assert_json_snapshot;

    #[test]
    fn test_user_create_request() {
        assert_json_snapshot!(UserCreateRequest {
            email: "test@example.com".to_string(),
            password: "12345678".to_string(),
        });
    }

    #[test]
    fn test_user_create_response() {
        assert_json_snapshot!(UserCreateResponse {
            email: "test@example.com".to_string(),
        });
    }

    #[test]
    fn test_session_create_request() {
        assert_json_snapshot!(SessionCreateRequest {
            email: "test@example.com".to_string(),
            password: "12345678".to_string(),
        });
    }

    #[test]
    fn test_session_create_response() {
        assert_json_snapshot!(SessionCreateResponse {
            token: "some-magik-token".to_string(),
        });
    }

    #[test]
    fn test_quote_file_create_request() {
        assert_json_snapshot!(QuoteFileCreateRequest {
            name: "Good quotes".to_string(),
        });
    }

    #[test]
    fn test_quote_file_create_response() {
        assert_json_snapshot!(QuoteFileCreateResponse {
            file: QuoteFile {
                id: 1,
                name: "Good quotes".to_string(),
                access: Permission::Manage,
            }
        });
    }

    #[test]
    fn test_quote_file_list_response() {
        assert_json_snapshot!(QuoteFileListResponse {
            files: vec![
                QuoteFile {
                    id: 1,
                    name: "Good quotes".to_string(),
                    access: Permission::Manage,
                },
                QuoteFile {
                    id: 2,
                    name: "Bad quotes".to_string(),
                    access: Permission::Read,
                }
            ],
        });
    }

    #[test]
    fn test_quote_file() {
        assert_json_snapshot!(QuoteFile {
            id: 1,
            name: "Good quotes".to_string(),
            access: Permission::Manage,
        });
    }

    #[test]
    fn test_permission() {
        assert_json_snapshot!(Permission::Manage);
        assert_json_snapshot!(Permission::Write);
        assert_json_snapshot!(Permission::Read);
    }

}
