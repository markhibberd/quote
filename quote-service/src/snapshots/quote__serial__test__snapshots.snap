---
source: src/serial.rs
expression: value
---
[
    1,
    2,
    3,
]
